handler.go

```diff
@@ -0,0 +1,45 @@
+package fosite
+
+import (
+	"errors"
+	"golang.org/x/net/context"
+	"net/http"
+)
+
+var ErrInvalidResponseType = errors.New("This handler is unable handle any of the response types requested by the auhtorize request")
+var ErrNoResponseTypeHandlerFound = errors.New("None of the handler's are able to handle this authorize request")
+
+type ResponseTypeHandler interface {
+	// HandleResponseType handles an authorize request. To extend the handler's capabilities, the http request
+	// is passed along, if further information retrieval is required.
+	//
+	// If HandleResponseType fails, the handler implementation MUST return ErrInvalidResponseType.
+	HandleResponseType(context.Context, *Response, AuthorizeRequest, http.Request) error
+}
+
+// NewAuthorizeResponse iterates through all response type handlers and returns their result or
+// ErrNoResponseTypeHandlerFound if none of the handler's were able to handle it.
+//
+// Important: Every ResponseTypeHandler should return ErrInvalidResponseType if it is unable to handle
+// the given request and an arbitrary error if an error occurred
+func (o *OAuth2) NewAuthorizeResponse(ctx context.Context, ar *AuthorizeRequest, r *http.Request) (*Response, error) {
+	var resp = new(Response)
+	var err error
+	var found bool
+
+	for _, h := range o.ResponseTypeHandlers {
+		// Dereference http request and authorize request so handler's can't mess with it.
+		err = h.HandleResponseType(ctx, resp, *ar, *r)
```

Comment:
NewAuthorizeResponse iterates through all ResponseTypeHandlers defined in the configuration. Each handler has access to the response and is able to modify the response. For example, the TokenResponseTypeHandler would initiate the [implicit grant](https://tools.ietf.org/html/rfc6749#section-4.2.1) and return an *access_token* instead of an *code*.

I am not sure how conflicts in the response's output keys should work. Should it be allowed? Should it throw an error? Should the response types decide? Example: Both AResponseTypeHandler sets `response.Set("foo", "bar")` and BResponseTypeHandler sets `response.Set("foo", "baz")`.